service/history/workflow/mutable_state_impl.go

```diff
@@ -2914,7 +2946,18 @@ func (ms *MutableStateImpl) addBuildIdToLoadedSearchAttribute(
 	if !stamp.GetUseVersioning() && effectiveBehavior == enumspb.VERSIONING_BEHAVIOR_UNSPECIFIED { // unversioned workflows may still have non-nil deployment, so we don't check deployment
 		newValues = append(newValues, worker_versioning.UnversionedSearchAttribute)
 	} else if effectiveBehavior == enumspb.VERSIONING_BEHAVIOR_PINNED {
-		newValues = append(newValues, worker_versioning.PinnedBuildIdSearchAttribute(ms.getPinnedDeployment()))
+		vInfo := ms.GetExecutionInfo().GetVersioningInfo()
+		if vInfo.GetVersioningOverride().GetPinnedVersion() != "" {
+			if vInfo.GetVersion() != "" { // worker deployments
+				newValues = append(newValues, worker_versioning.PinnedBuildIdSearchAttribute(nil, vInfo.GetVersioningOverride().GetPinnedVersion()))
```

Comment:
similar to `assigned:<bld>` and `unversioned`, we only want one `pinned:...` value in the list. old ones should be added as ordinary values (I guess they still have a `versioned:` or `unversioned:` prefixes).

Reply:
ah ok, I think I didn't handle that correctly in the last release. Let me refactor this ..